FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java

```diff
@@ -205,7 +205,7 @@ protected void layoutChildren(double x, double y, double w, double h) {
 
         rightButton.resize(clearButtonWidth, h);
         positionInArea(rightButton,
-                (x+w) - clearButtonWidth, y,
+                (x+w) + textField.snappedLeftInset() + textField.snappedRightInset() - (h + textField.snappedTopInset() + textField.snappedBottomInset()), y,
```

Comment:
We need the right inset because we set it to a higher value, so the text no longer overlaps. Otherwise it would look like this:
![noinsets](https://user-images.githubusercontent.com/18708736/115241914-a50cbc00-a121-11eb-87d7-b66c51eb6b60.png)
Explanation:
![explanation](https://user-images.githubusercontent.com/18708736/115242420-22d0c780-a122-11eb-9032-72dcd99457d5.png)
We want to calculate the x position of z. The formula for this is `textfield_x + full_width_textfield - full height-textfield`. I would also not remove any other insets from the calculation because if another user sets custom insets the button may be positioned wrongly.

Reply:
The way i thought:
The button should be a square (=> `width = height`). The easiest way to calculate x position of the button is to take the entire width of the textfield and subtract the width. In this case `width = height` so the x position of the button is `xpos_button = xpos_textfield + width_textfield - width_button = xpos_textfield + width_textfield - height_button`

Reply:
It could be rectangular. It depends on the dimensions the programmer defines for the whole textfield. In any way the button would need to cover the height of the textfield, so could end up being rectangular (i.e. non square)

Reply:
I don't quite get that. If i set the width and the height of the button to the same value it must be a square.

Reply:
I mean.. this is a general purpose library for adding better skins to existing apps. 
The programmer might decide to make textfields height taller by changing their padding values in that case the button should grow to fill the textfield height making it possible for the button to be rectangular (not square).
The idea here would be for the button to cover the textfield height, not for it to be restricted to being a square, I think.

Reply:
I have a better solution in the working that should fix the issues. But i encountered the following issue:
`clearButtonWidth` does not actually calculate the width of the graphic. It always returns the same value.
I could bypass this by adding a minimum width to the right button. This would also lead to the nice sideeffect that the textfield right-button and the password-textfield right-button have the same width (if possible due to textfield dimensions).

Reply:
Current code is as follows:
```java
final double clearGraphicWidth = snapSizeX(rightButtonGraphic.prefWidth(-1));
final double clearButtonWidth = rightButton.snappedLeftInset() + clearGraphicWidth + rightButton.snappedRightInset();
        
rightButton.resize(clearButtonWidth, textField.getHeight());
positionInArea(rightButton, textField.getWidth() - clearButtonWidth, 0, clearButtonWidth, textField.getHeight(), 0, HPos.LEFT, VPos.TOP);
```

I still need to adjust the width according to the textfield dimensions

Reply:
Have you checked the comment I posted for the possible size, x and y definitions for the right button?
```
x = (x + w - textField.snappedRightInset()) - clearButtonWidth;
y = y + textField.snappedTopInset()
width = clearButtonWidth
height = h - (textField.snappedTopInset() + textField.snappedBottomInset())
```

The `prefWidth` method should give you the width of the right button according to what's defined in the css, etc..

Some comments to the code of your last comment that you posted:
1. I think you forgot to factor in the padding of the textfield (i.e. the insets). 
2. You don't need to get the height of the textfield: `textfield.getHeight()` and the width, they are already passed on to `layoutChildren` method.
3. You're resizing the button twice by calling `resize` and then `positionInArea` (`positionInArea` resizes and sets the x and y position).

Reply:
I tested your code and it does not place the button correct at all. `prefWidth` returns `0` for the password-textfield and `10.4` is returned for the normal textfield.

Regarding 1+2:
Printout for a simple textfield: `h`: 17.6 `textfield.getHeight()`: 25.6
=> `h != textfield.getHeight()`. I think that `textfield.getHeight()` takes the padding/insets into account and `h` does not

Regarding 3:
If we do not resize the button it will look like this:
![grafik](https://user-images.githubusercontent.com/18708736/117131335-9a008f80-ada1-11eb-84e9-a7b10e387f03.png)

Reply:
1 - There must be something else happening about `prefWidth` not working. I've been working with JavaFX since it's very first release (about 10/15 years ago) non stop, I've always relied on `prefWidth`, `prefHeight` for layout code. It should return a valid value unless there is something wrong with the size definitions we are setting for it in css, etc or something else.
One thing we could change when calling `rightButtonGraphic.prefWidth(-1)` is pass in the height, i.e: `rightButtonGraphic.prefWidth(h)` since we already know the height we want for the right button (we only pass -1 when we don't know the height).

2 - Yeah I think you're right., height and width passed might already be factoring in the padding of the textfield. 
Anyways, we don't need to be calling textfield.getHeight(), etc we can just use the width, height passed in to the layoutChildren method which are the dimension we should be using for laying out our nodes (they are the dimensions of the area that is provided for us to layout our nodes in the layoutChildren method).

3 - `resize`, sets the button width and height. `positionInArea` does a similar job: resizes and positions. 
If you think there is some issue with `positionInArea`, we could call `resize` (to set width and height) and then `setLayoutX` and `setLayoutY` (to position in the x and y axis). So that we don't repeat the same computation..

Reply:
Regarding 3: `resize` was used even before my PR

Reply:
3 - Yeap, you're right. My mistake, not yours, I don't think I should have been calling resize and positionInArea, I don't think that makes logical sense? As I was repeating the same computation twice.

(some of this code is quite old, JMetro has been around since 2013 if I'm not mistaken, some of it might be outdated or wrong as through the years I've learned more about JavaFX, etc)

Anyways sorry to bother you with some of my past mistakes. I think as we're re-implementing this code we might as well fix some of this mistakes.

Reply:
So you are saying i should somehow get rid of `resize`

Reply:
You can call `resize` and then `setLayoutX` and `setLayoutY ` that way we don't repeat the "resize" part.
There's also the `resizeRelocate(...)` method, that we could use, which does both things in one go (setting width, height and x and y).

Reply:
The only thing i can imagine that could cause `prefWidth` to return 0 is `-fx-scale-shape: false;` for the password-textfield. The is that we cant remove it because otherwise the graphic is not being drawn (or scaled very wrongly).

Reply:
Currently im using the padding to adjust the width of the button (which is configurable by the user)

Reply:
I just tried calling in your code in the `layoutChildren` method: 
``` 
double rightButtonWidth = rightButton.prefWidth(-1);
System.out.println("Right button width = " + rightButtonWidth);
```
And it prints `36`

There might be some detail that escaped you...?

However, since we know the height I think we should actually calling it this way:
```
double rightButtonWidth = rightButton.prefWidth(h);
```

Reply:
You are right with `rightButton.prefWidth(-1)` but i meant `rightButtonGraphic.prefWidth(-1)`.
```
TextField:
Right button width = 36.4
Right button graphic width = 10.0
password-textfield:
Right button width = 26.0
Right button graphic width = 0.0
```
The problem (at least i think) is that `rightButton.prefWidth` depends on `rightButtonGraphic.prefWidth` which returns DIFFERENT values for `TextField` (X-Icon) and `password-textfield` (Eye-Icon).

Reply:
`rightButton.prefWidth(h)` returns the following (when the padding is removed again):

```
TextField: 10.4
password-textfield: 0.0
```

They should be the same

Reply:
It works if i do `rightButton.setPrefWidth` in `layoutChildren`

Reply:
The right button graphic is returning `0` for the passwordfield because we haven't defined a padding value for the passwordfield right button graphic in CSS.

Reply:
> It works if i do `rightButton.setPrefWidth` in `layoutChildren`

It doesn't sound right to set a `prefWidth` in the `layoutChildren` method. 
The right button graphic is returning 0 for the passwordfield because we haven't defined a padding value for the passwordfield right button graphic in CSS.

If we want the width of the right button to be exactly the same for the passwordfield and the textfield then we can play with setting padding values on rightButton/rightButtonGraphic in CSS for the passwordfield and textfield.
It also has the added bonus of programmers being able to define whatever width they want for the right button by defining their own css rules.

Reply:
But the problem is that if i add padding the graphic scales differently

How its supposed to look:
![grafik](https://user-images.githubusercontent.com/18708736/117175016-a8b16b80-adce-11eb-9562-502fe9548a26.png)

With increased padding (from 0.416667 to 0.8em):
![grafik](https://user-images.githubusercontent.com/18708736/117175140-cbdc1b00-adce-11eb-8774-b35c4986124f.png)

With increased padding (from 0.416667 to 0.8em) and `-fx-scale-shape: false` (just like the `password-textfield` icon):
![grafik](https://user-images.githubusercontent.com/18708736/117175176-d8607380-adce-11eb-9677-0345cbc1a5de.png)

Reply:
```
-fx-scale-x: 0.65;
-fx-scale-y: 0.65;
-fx-scale-z: 0.65;
```
would fix it actually

Reply:
I wasn't saying to increase the padding in the right button graphic of the textfield. We can leave as it is..
We can adjust the padding on the right button instead so that the sum of the right button with the rightbutton graphic padding gets you the width you want.

Reply:
Thanks for the suggestion of using `-fx-scale-x`
Though I think we could just use padding instead, makes it simpler. We can play with the padding of the right button and with the padding of the right button graphic.

Reply:
But how does the graphic actually behave for high-dpi displays?

Reply:
I think we're probably not on the same page.
I probably didn't explain myself well.

On your previous comment you said you tried to change the padding on the right button graphic for the cross button. That was not what I suggested.
I'm saying leave the padding of the right button graphic for the cross button as it is but adjust the padding on the right button (not the right button graphic) if we want to make the right button width similar to the right button width of the password field. I think that's what you wanted and what we are now discussing here?

Reply:
I realised that and i committed a corrected version yesterday. See 756a7ed

Reply:
Great! Thanks and thanks for your patience in discussing these issues.

Just one tiny small last suggestion before I approve and merge. Could we have it like this, instead of calling `textField.prefWidth(-1)` since we already know the right button height:

`layoutChiildren` method:
```
final double rightButtonHeight = h + textField.snappedTopInset() + textField.snappedBottomInset();
final double rightButtonWidth = rightButton.prefWidth(rightButtonHeight);
```

Reply:
I don't know what you mean. We never call `textField.prefWidth(-1)`

Reply:
I already changed that in 1cd80a4

Reply:
Yeah sorry, I meant `rightButton.prefWidth(-1);` 
( I meant to write `rightButton` but I wrote textField instead).